debounce/simple_debouncer_test.go

```diff
@@ -5,17 +5,154 @@ package debounce
 
 import (
 	"runtime"
+	"sync"
 	"sync/atomic"
 	"testing"
+	"time"
 )
 
+// TestSimpleDebouncerRace tests SimpleDebouncer for the fact that it does not allow concurrent writing, reading.
+func TestSimpleDebouncerRace(t *testing.T) {
+	operations := 1000
+	runs := 100
+	count := 3
+
+	d := NewSimpleDebouncer()
+	for r := 0; r < runs; r++ {
+		var counter atomic.Int32
+		var wg sync.WaitGroup
+		wg.Add(count)
+
+		results := make([]bool, count)
+		fails := make([]bool, count)
+		for c := range results {
+			result := &results[c]
+			fail := &fails[c]
+
+			go func() {
+				*result = d.Debounce(func() {
+					for i := 0; i < operations; i++ {
+						if counter.Add(1) != 1 {
+							*fail = true
+						}
+						time.Sleep(time.Microsecond)
+						counter.Add(-1)
+					}
+				})
+				wg.Done()
+			}()
+		}
+		wg.Wait()
+
+		// check results
+
+		finished := 0
+		for i, done := range results {
+			if done {
+				finished++
+			}
+			if fails[i] {
+				t.Fatalf("Simultaneous execution detected")
+			}
+		}
+		if finished < 2 {
+			t.Fatalf("In one run should be finished more than 2 `Debounce` method calls, but finished %d", finished)
+		}
+	}
+}
+
+// TestDebouncerExtreme tests SimpleDebouncer in the conditions  fast multi `Debounce` method calls and fast execution of the `debounced function`.
+func TestDebouncerExtreme(t *testing.T) {
+	type runResult struct {
+		executedN int32
+		done      bool
+	}
+
+	runs := 10000
+	count := 20
+
+	d := NewSimpleDebouncer()
+	var wg sync.WaitGroup
+	for r := 0; r < runs; r++ {
+		var executionsC atomic.Int32
+		wg.Add(count)
+
+		results := make([]runResult, count)
+
+		for c := range results {
+			result := &results[c]
+
+			go func() {
+				result.done = d.Debounce(func() {
+					result.executedN = executionsC.Add(1)
+				})
+				wg.Done()
+			}()
+		}
+		wg.Wait()
+
+		// check results
+		finished := 0
+		for _, result := range results {
+			if result.done {
+				if result.executedN == 0 {
+					t.Fatalf("Wrong execution detected: \n%#v", result)
+				}
+				finished++
+			}
+		}
+		if finished < 2 {
+			t.Fatalf("In one run should be finished more than 2 `Debounce` method calls, but finished %d", finished)
+		}
+	}
+}
+
+// TestSimpleDebouncerCount tests SimpleDebouncer for the fact that it pended only one function call.
+func TestSimpleDebouncerCount(t *testing.T) {
```

Comment:
This test tests a thing that is already tested in `TestSimpleDebouncerRace`, what is the reason for it ?

Reply:
The purpose of the `TestSimpleDebouncerRace` and `TestSimpleDebouncerCount` tests are different.
The predication and logic of `TestSimpleDebouncerRace` does not imply a thorough calculation of the calls made, although it is carried out there, but only for the purpose of determining that there were simultaneous calls.
The predication and logic of `TestSimpleDebouncerCount` assumes a specific count of completed calls.
They can be combined, but then their logic will be more confusing due to the combination of different purposes.
And this is reflected in the verification conditions:
```
func TestSimpleDebouncerRace(t *testing.T) {
...
	if finished < 2 {
			t.Fatalf("In one run should be finished more than 2 `Debounce` method calls, but finished %d", finished)
		}
```
and 
```
func TestSimpleDebouncerCount(t *testing.T) {
...
	if finished != 2 {
		t.Fatalf("Should be finished 2 `Debounce` method calls, but finished %d", finished)
	}
```
For more clarity, the`TestSimpleDebouncerCount` error messages have been edited:
```
	t.Fatalf("Should be finished 2 `Debounce` method calls, but finished %d", finished)
```